### PR TITLE
Array expected value badges

### PR DESCRIPTION
**Pull request by luisbettencourt**
## 🌟 What is the purpose of this PR?

This PR adds the missing array badge types and displays them on the expected values column on the properties table of the entity type page.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203363157432094/1203490593648606/f) _(internal)_

## 🚫 Blocked by

- [ ] #1769 

## 🔍 What does this change?

- Added the `Text Array`, `Boolean Array` and `Number Array` badges;
- Changed the type of arrays with 2 or more different data types to be `Mixed Array` instead of `DataType Array`;
- Simplified `getArrayExpectedValueType` to make it easier to use for the expected values column in the properties table;
- Added an icon next to the property's title whenever `Allow arrays` is selected;
- Minor styling changes around the property title and expected value chips;
- Changed the `ExpectedValueSelector` Autocomplete to `freeSolo` to prevent the warning `The value provided to Autocomplete is invalid`.

## 📹 Demo

image

image

### CODE REVIEW

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-to-form-data-expected-values.ts [hunk 1]
```diff
@@ -41,6 +41,7 @@ export const propertyTypeToFormDataExpectedValues = (
           data: {
             typeId: expectedValue.$ref,
           },
+          ...(parentId ? { parentId } : {}),
```
Comment (luisbettencourt) [inline packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-to-form-data-expected-values.ts:44]: This fix should be made in #1769, just added it here for testing purposes

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx [hunk 2]
```diff
@@ -1,36 +1,58 @@
...
-  Record<string, string>
->((prev, dataType) => {
-  const { dataTypeId, title } = dataType;
-
-  return { ...prev, [dataTypeId]: title };
-}, {});
+import { expectedValuesOptions } from "./shared/expected-values-options";
+import { getArrayExpectedValueType } from "./shared/get-expected-value-descriptor";
 
 // @todo handle this being too many
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx:9]: This you've handled this todo?

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx [hunk 3]
```diff
@@ -1,36 +1,58 @@
...
+        expectedValueOption = expectedValuesOptions[arrayType];
       }
 
-      if (label) {
-        return <Chip key={label} label={label} color="gray" />;
+      if (expectedValueOption) {
+        return (
+          <Chip
+            // eslint-disable-next-line react/no-array-index-key
+            key={`${property.$id}-${index}`}
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx:35]: There's little benefit to prepending index with anything in particular here

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-form/expected-value-selector.tsx [hunk 4]
```diff
@@ -163,7 +175,8 @@ const ExpectedValueSelector: ForwardRefRenderFunction<
             return (
               <ExpectedValueChip
                 {...getTagProps({ index })}
-                key={typeId}
+                // eslint-disable-next-line react/no-array-index-key
+                key={index}
```
Comment (nathggns) [inline packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-form/expected-value-selector.tsx:179]: You can probably use something like `typeof expectedValue === "object" ? expectedValue.id : expectedValue` here

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-to-form-data-expected-values.ts (hunk 1 above)
Comment (nathggns) [reply packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-to-form-data-expected-values.ts:44]: Been added in that PR now

**Review by nathggns — COMMENTED** [review]
Just comments for now as conflicts – will re-review later

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx (hunk 2 above)
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx:9]: Not sure what the comment is about. You added it, do you remember what it was?

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx (hunk 3 above)
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx]: Removed the prefix in 7299c6b6ecea7fb0fb61a0a82839fe50c5594c14

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-form/expected-value-selector.tsx (hunk 4 above)
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-type-form/expected-value-selector.tsx]: Thank you, followed your suggestion in 2ad67387d9c8f0453ecf2213867beda796492e5e

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/shared/get-expected-value-descriptor.ts [hunk 5]
```diff
@@ -53,7 +62,11 @@ export const getExpectedValueDescriptor = (
     case "array":
       return {
         typeId: "array",
-        arrayType: getArrayExpectedValueType(data, flattenedExpectedValues),
+        arrayType: getArrayExpectedValueType(
+          Object.values(flattenedExpectedValues)
+            .filter(({ parentId }) => parentId === id)
+            .map(({ data: childData }) => childData!.typeId),
```
[the remark this answers is not in the document]
Comment (luisbettencourt) [reply packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/shared/get-expected-value-descriptor.ts]: Added in a7571ea6c2953b80831c0734b036fab028eb93bb

File: packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx (hunk 2 above)
Comment (nathggns) [reply packages/hash/frontend/src/pages/[shortname]/types/entity-type/[entity-type-id].page/definition-tab/property-list-card/property-expected-values.tsx:9]: It was just about it appearing properly when there were too many types to fit on one row, which I think you've done now

**Review by nathggns — DISMISSED** [review]

**Pull request merged**